Index refactor

rjgildea opened the pull request:
1. Refactor basis vector search code into strategies in `dials.algorithms.indexing.basis_vector_search.strategies`
    - fft1d, fft3d, real_space_grid_search
    - take as input a list of reciprocal lattice vectors and return a list of candidate basis vectors
2. Refactor basis vector combination code into `algorithms.indexing.basis_vector_search.combinations`
    - Implemented as a sequence of iterators that can be chained together
    - candidate_orientation_matrices: finds suitable combinations of basis vectors in P1
    - filter_known_symmetry: filters output of previous to only accept those models that match known symmetry
    - filter_similar_orientations
3. Refactor assign_indices to a strategy pattern in `dials.algorithms.indexing.assign_indices`
    - assign_indices_global, assign_indices_local
4. Move non-primitive basis code to separate module `dials.algorithms.indexing.non_primitive_basis` with tests
5. New `dials.algorithms.indexing.model_evaluation`
    - Move/rename SolutionTrackers: `ModelRankWeighted`, `ModelRankFilter`
    - move run_one_refinement inner function to new `ModelEvaluation` class
6. Attempt to encapsulate handling of known symmetry into external class `dials.algorithms.indexing.symmetry.SymmetryHandler`, with tests
7. New `dials.algorithms.indexing.lattice_search` module
    - Single `BasisVectorSearch` inheriting from `indexer_base` to consolidate previous `fft1d`, `fft3d` and `real_space_grid_search` classes. Code that is specific to a basis vector search now lives in this class instead of `indexer_base`.
    - Add simple test for `BasisVectorSearch` class
8. Move basis vector optimisation code to new module `dials.algorithms.indexing.basis_vector_search.optimise` with tests
9. Replace scan_range parameter with image_range
10. Refactor command line script to a functional interface
    - add joint_indexing=False option to allow indexing multiple independent experiments in a single dials.index process
11. Some refactoring of indexer.index() method to minimise code duplication with stills_indexer

Commit:
refactor basis vector search code into strategies

Commit:
refactor basis vector combination code

Commit:
use basis vector search strategies

Commit:
use refactored basis vector combination code

Commit:
skip these tests as cluster_analysis_search=True not currently implemented

Commit:
minor relaxation of tolerance

Commit:
refactor assign_indices to strategy pattern; move to separate file; move test location; enblacken

Commit:
find_basis_vectors also returns used_in_indexing selection: enables indirectly setting d_min for initial refinement

Commit:
move filtering of similar unit cell orientations to outside of choose_best_orientation_matrix()

Commit:
missing import

Commit:
move non-primitive basis code to separate module; add tests

Commit:
use new location for non_primitive_basis.correct()

Commit:
move/rename SolutionTrackers to new model_evaluation module

Commit:
move run_one_refinement inner function to new class in model_evaluation

Commit:
pytest.approx

Commit:
this code now lives in model_evaluation

Commit:
Attempt to encapsulate handling of known symmetry into external class

Commit:
Consolidate fft1d, fft3d, rsgs classes into single BasisVectorSearch class
inheriting from indexer_base

Commit:
Add test for stills indexer; fix stills indexer

Commit:
this code has already been moved elsewhere

Commit:
Add reflection_table.calculate_entering_flags()

Replaces duplicated code in indexing/refinement.

Commit:
Move basis vector optimisation code to new module.

- reinstate optimise_initial_basis_vectors option
- move shared pytest fixtrue to conftest

Commit:
remove unused functions

Commit:
Refactor dials.index command line script.

- move towards a "functional interface" (#691)
- replace scan_range parameter with image_range

Commit:
reinstate max_combinations parameter

Commit:
Move phil definitions closer to code where they are used

Commit:
scan_range -> image_range

Commit:
remove non-existing import

Commit:
pickle -> msgpack

Commit:
fix test

Commit:
move all indexing tests to algorithms/indexing

Commit:
remove deprecated show_rotation_matrix_differences() function

Commit:
add test for rotation_matrix_differences() function

Commit:
Remove commented code; comments-> docstring

Commit:
improve code coverage of tests

Commit:
remove dead code

Commit:
fix import after moving test location

Commit:
soften tests for optimise_basis_vectors

Commit:
remove debug_plots functionality

Commit:
improve code coverage of max_cell tests

Commit:
remove debug_plot_candidate_basis_vectors() function

Commit:
add simple test for BasisVectorSearch class

Commit:
start refactoring of indexer.index() function

Commit:
Add stills indexer BasisVectorSearch test

Commit:
these files should have been removed with fed639b39345782ff833ba89fb928554c9e4f887

Commit:
move filter_similar_orientations to standalone function; add test

Commit:
fix tests

Commit:
docstrings for strategies module

Commit:
search_beam_position: scan_range -> image_range; add test

Commit:
add option to index multiple experimemts independently (joint_indexing=False)

Commit:
mark method as private

Commit:
updated index api docs

Commit:
test for indexing multiple experiments with joint_indexing=False

Commit:
don't use locals()

jbeilstenedmands committed:
Fix scaling bug that can occur when some bin stats are None

jbeilstenedmands committed:
Remove option to specify space group in dials.scale

This is not currently well handled, and should instead be
handled with dials.reindex or symmetry/cosym

DiamondLightSource-build-server committed:
Clean clutter

Please be advised that you can now install pre-commits by running

    libtbx.precommit install

This will fix up the code for you and avoids clutter getting into
the repository in the first place.

jbeilstenedmands committed:
Remove custom pytest marker from dials.scale tests

ndevenish committed:
Don't set matplotlib backend in per_image_analysis

This directly sets the Agg canvas instead of relying on global state.

This fixes #795 by still allowing GUI plots if required by the user.

ndevenish committed:
Flake8 and tidy per_image_analysis.py

- Flake8
- Float imports
- Sort imports

ndevenish committed:
Add missed 'Sorry' imports

ndevenish commented:
Checking with quality tools; there are some outright undefined variables; I checked the `stills_indexer` ones and these were moved out in this branch. The others might have existed anyway. This is obviously not covered by tests.
```
algorithms/indexing/model_evaluation.py:53:42: F821 undefined name 'difference_rotation_matrix_axis_angle'
algorithms/indexing/stills_indexer.py:118:20: F821 undefined name 'stills_indexer_fft3d'
algorithms/indexing/stills_indexer.py:120:20: F821 undefined name 'stills_indexer_fft1d'
algorithms/indexing/stills_indexer.py:122:20: F821 undefined name 'stills_indexer_real_space_grid_search'
```

Otherwise, there is a whole load of `flake8` unused variables; I know not all these files have been completely rewritten but it'd be nice to get rid of the flake8 warnings; most of these are probably harmless, but I'm worried about cases where the variable should be used but isn't - is impossible for me to tell the difference...

```
algorithms/indexing/assign_indices.py:40:9: F841 local variable 'diffs' is assigned to but never used
algorithms/indexing/assign_indices.py:41:9: F841 local variable 'norms' is assigned to but never used
algorithms/indexing/assign_indices.py:42:9: F841 local variable 'hkl_ints' is assigned to but never used
algorithms/indexing/assign_indices.py:112:9: F841 local variable 'diffs' is assigned to but never used
algorithms/indexing/assign_indices.py:113:9: F841 local variable 'norms' is assigned to but never used
algorithms/indexing/assign_indices.py:114:9: F841 local variable 'hkl_ints' is assigned to but never used
algorithms/indexing/assign_indices.py:141:13: F841 local variable 'ref_sel' is assigned to but never used
algorithms/indexing/basis_vector_search/combinations.py:74:9: F841 local variable 'best_model' is assigned to but never used
algorithms/indexing/basis_vector_search/conftest.py:22:5: F841 local variable 'sg' is assigned to but never used
algorithms/indexing/basis_vector_search/strategies.py:267:9: F841 local variable 'sites_frac' is assigned to but never used
algorithms/indexing/basis_vector_search/strategies.py:282:9: F841 local variable 'orth' is assigned to but never used
algorithms/indexing/basis_vector_search/strategies.py:390:9: F841 local variable 'rlgrid' is assigned to but never used
algorithms/indexing/basis_vector_search/test_combinations.py:69:5: F841 local variable 'crystal_symmetry' is assigned to but never used
algorithms/indexing/indexer.py:831:17: F841 local variable 'isel' is assigned to but never used
algorithms/indexing/indexer.py:832:17: F841 local variable 'ref_panel' is assigned to but never used
algorithms/indexing/indexer.py:922:9: F841 local variable 'verbosity' is assigned to but never used
algorithms/indexing/model_evaluation.py:343:9: F841 local variable 'e' is assigned to but never used
algorithms/indexing/nearest_neighbor.py:96:9: F841 local variable 'centers' is assigned to but never used
algorithms/indexing/nearest_neighbor.py:138:9: F841 local variable 'fig' is assigned to but never used
algorithms/indexing/non_primitive_basis.py:30:9: F841 local variable 'cryst_orig' is assigned to but never used
algorithms/indexing/refinement.py:21:5: F841 local variable 'detector' is assigned to but never used
algorithms/indexing/stills_indexer.py:55:9: F841 local variable 'PX' is assigned to but never used
algorithms/indexing/stills_indexer.py:56:9: F841 local variable 'fig' is assigned to but never used
algorithms/indexing/stills_indexer.py:92:5: F841 local variable 'history' is assigned to but never used
algorithms/indexing/stills_indexer.py:590:21: F841 local variable 'crystal_model_nv0' is assigned to but never used
algorithms/indexing/stills_indexer.py:793:9: F841 local variable 'crystal_model_nv0' is assigned to but never used
algorithms/indexing/symmetry.py:161:5: F841 local variable 'supergroup' is assigned to but never used
algorithms/indexing/symmetry.py:163:5: F841 local variable 'triclinic_miller' is assigned to but never used
algorithms/indexing/symmetry.py:622:9: F841 local variable 'primitive_subsym' is assigned to but never used
algorithms/indexing/test_assign_indices.py:80:5: F841 local variable 'miller_set' is assigned to but never used
algorithms/indexing/test_index.py:52:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:120:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:151:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:226:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:263:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:300:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:337:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:372:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:403:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:438:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:476:9: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:515:13: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:563:5: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:607:5: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:633:5: F841 local variable 'result' is assigned to but never used
algorithms/indexing/test_index.py:777:5: F841 local variable 'experiments_list' is assigned to but never used
algorithms/indexing/test_lattice_search.py:58:5: F841 local variable 'indexed_reflections' is assigned to but never used
algorithms/indexing/test_lattice_search.py:99:5: F841 local variable 'indexed_reflections' is assigned to but never used
algorithms/indexing/test_max_cell.py:25:5: F841 local variable 'expts' is assigned to but never used
algorithms/indexing/test_symmetry.py:74:5: F841 local variable 'cs_new' is assigned to but never used
algorithms/indexing/test_symmetry.py:120:5: F841 local variable 'subgroups' is assigned to but never used
```

Anthchirp committed:
Remove setpaths* from installer

Forces users to run the installer instead of running code in place.
Resolves SCI-8659

Remove precommitbx environment

Commit:
missing import; add test

Commit:
I don't think this method is called

Commit:
address flake8 unassigned local variable warning (F841)

Commit:
fix test_compare_orientation_matrices

- if rotation angle is small, rotation axis is unstable/irrelvant
- workaround -0.000 vs 0.000 in text output

Commit:
fix test_phi_scan indavertantly broken by 4f9f801fdbfcbac89197ecfa129befdb1ec32802

rjgildea commented:
@ndevenish have addressed the flake8 issues highlighted above

Anthchirp committed:
This is true by default.

Fixes flake8 warning in per-image-analysis pull request

Commit:
replace Sorry with custom DialsIndexError

Commit:
loosen test tolerance

Commit:
catch DialsIndexError and re-raise as Sorry

jbeilstenedmands committed:
Replace some Sorry exceptions in scaling.

Commit:
Parallel indexing of multiple experiments

- adds dependency on python2.7 backport of concurrent.futures

Commit:
PEP8 class naming convention

Commit:
pacify flake8

Commit:
tidy imports

Commit:
I don't think this is required any more

jbeilstenedmands committed:
Replace Sorry calls with ValueError where appropriate.

Modify command line scripts which call these to catch and throw
a Sorry.

ndevenish committed:
Restore reflection selection in reciprocal_lattice_viewer

Was missed in refactor in 5def7eb893f2e4a53da1dbca93e79e3c21f693d7

Fixes #806

Commit:
fix some lgtm issues

Commit:
RLV: imagesets -> experiments

Commit:
>= not >

Anthchirp committed:
Add .flake8 file for manual validation

disables checks for things managed by black

Anthchirp committed:
Fix dials.rl_png command

due to recent changes in RLV. Some campsite improvements.

Anthchirp committed:
fix spacing in error message

jbeilstenedmands committed:
Fix assign_identifiers bug introduced in #7383e9a

Anthchirp committed:
Fix dials.export format=json

due to recent changes in RLV

jbeilstenedmands committed:
Remove use of sorry from non-cmd export code.

Tidy up some bits of export too.

Commit:
add deprecated index_reflections() function for backwards compatibility

Commit:
this break statement can never be hit

Commit:
remove unreachable code

Anthchirp committed:
precommitbx: build libffi

and update OpenSSL

Commit:
xrange -> range

Commit:
perhaps this isn't needed after all?

Commit:
missed off unindexed reflections (removed by 667b8c8e985e7b5b9bbf167eda33221a494e4f86)

Commit:
fix find_spots_server

Commit:
fix broken test logic

Commit:
fix flake8 issues

Commit:
remove unused variable

Commit:
don't convert generator to list as might not need all items

Commit:
implement basis vector search strategies as entry points

Commit:
print -> logging

Commit:
Appease the flake8 Gods
